driver-core/src/main/com/mongodb/internal/async/function/RetryState.java

```diff
@@ -0,0 +1,379 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.internal.async.function;
+
+import com.mongodb.annotations.NotThreadSafe;
+import com.mongodb.internal.async.SingleResultCallback;
+import com.mongodb.internal.async.function.LoopState.AttachmentKey;
+import com.mongodb.lang.NonNull;
+import com.mongodb.lang.Nullable;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+import java.util.function.BiPredicate;
+import java.util.function.Supplier;
+
+import static com.mongodb.assertions.Assertions.assertFalse;
+import static com.mongodb.assertions.Assertions.assertNotNull;
+import static com.mongodb.assertions.Assertions.assertTrue;
+
+/**
+ * Represents both the state associated with a retryable activity and a handle that can be used to affect retrying, e.g.,
+ * to {@linkplain #breakAndThrowIfRetryAnd(Supplier) break} it.
+ * {@linkplain #attachment(AttachmentKey) Attachments} may be used by the associated retryable activity either
+ * to preserve a state between attempts.
+ *
+ * @see RetryingSyncSupplier
+ * @see RetryingAsyncCallbackSupplier
+ */
+@NotThreadSafe
+public final class RetryState {
```

Comment:
Lets add a unit test for this - so any refactoring in the future is simplified.

Reply:
Done.